forge/src/trace/decoder.rs

```diff
@@ -0,0 +1,234 @@
+use super::{
+    CallTraceArena, RawOrDecodedCall, RawOrDecodedLog, RawOrDecodedReturnData, TraceIdentifier,
+};
+use crate::abi::{CHEATCODE_ADDRESS, CONSOLE_ABI, HEVM_ABI};
+use ethers::{
+    abi::{Abi, Address, Event, Function, Token},
+    types::H256,
+};
+use foundry_utils::format_token;
+use std::collections::BTreeMap;
+
+/// The call trace decoder.
+///
+/// The decoder collects address labels and ABIs from any number of [TraceIdentifier]s, which it
+/// then uses to decode the call trace.
+///
+/// Note that a call trace decoder is required for each new set of traces, since addresses in
+/// different sets might overlap.
+#[derive(Default, Debug)]
+pub struct CallTraceDecoder {
+    /// Addresses identified to be a specific contract.
+    ///
+    /// The values are in the form `"<artifact>:<contract>"`.
+    pub contracts: BTreeMap<Address, String>,
+    /// Address labels
+    pub labels: BTreeMap<Address, String>,
+    /// All known functions
+    pub functions: BTreeMap<[u8; 4], Function>,
```

Comment:
We need to change this unfortunately... likely to `BTreeMap<[u8; 4], Vec<Function>>`, reason being, short signatures don't identify differences in output types. So if a user has `myFunc(uint256) external returns (uint256)` and just `myFunc(uint256) external`, the one added first will be overwritten and fail to decode.

we should make it a vector (or make the key different, but having the key be the same as the calldata input is nice), and iteratively attempt decoding. Vast majority of the time there will be only one but still.

Doesnt apply to events so that can stay the same

Reply:
Ah, right. That's.. annoying. I'm adding a test to https://github.com/onbjerg/foundry-test that checks for this behavior:

![image](https://user-images.githubusercontent.com/8862627/158029458-287f02de-cd92-4f6c-8a02-646b2f202155.png)

Reply:
I fixed it using a `BTreeMap<Address, BTreeMap<[u8; 4], Function>>` since the other solution (using a `Vec`) would have a lot of oddities, notably if a function returns `bytes` and another one `string` they could both decode the output so it would show up wrong (i.e. both might be decoded as `bytes` or `string`) :/

Reply:
ehhhhhhhhh - not sure we want to do this. this then *requires* us to have identified the contracts. When in actuality, if a sig matches, we should be able to decode it even if we have identified that specific contract (which can happen). I would really prefer to remain agnostic to the address and lean more on as better decoding scheme

Reply:
@brockelmore what do you have in mind? The `Vec` approach assumes there's a unique way to decode things as you iterate through the vector, which isn't right for the reason bjerg mentions above, although we could be opinionated and say "if it can show as ascii then go with string, else with bytes", but I suspect this can happen with any kind of fixed size return type, e.g. u256 and bytes32?

Keying by address I think might prevent us from using 4byte directory later on?

Reply:
1) it should be relatively rare. i accidentally did it awhile back and ended up just changed my contracts. most of the time i think ppl will do it by accident. gently nudging them to change just better decoding is probably a good idea. Maybe we leave as is (`BTreeMap<[u8;4], Function>`) and emit a warning?

Reply:
I am OK with leaving as-is and maybe doing a `trace::warn` (not logged by default), or not doing anything at all and waiting until a user bumps on it as a bug and decide then

Reply:
> Doesnt apply to events so that can stay the same

I'm not sure if this affects implementation here/anywhere, but just noting that events do have a similar problem but with `indexed` parameters.

If you have topic `0xd8c9334b1a9c2f9da342a0a2b32629c1a229b6445dad78947f674b44444a7550`, 4byte.directory tells you the sig is `event TextChanged(bytes32,string,string)`. However, that's not enough to decode the event because two of those paramers are indexed but we don't know which. You'd pretty much have to guess which params are indexed, and depending on the types it's possible multiple sigs may decode and you can't tell which is correct

Reply:
`event TextChanged(bytes32 indexed,string,string)` sig != `event TextChanged(bytes32,string,string)`, indexed params are included in the event signature calculation